File: JavaScript/lrucache.js
// 7. LRU Cache using Doubly Linked List + Map
class LRUNode {
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.prev = this.next = null;
    }
}

class LRUCache {
    constructor(capacity) {
        this.capacity = capacity;
        this.map = new Map();
        this.head = new LRUNode(0, 0);
        this.tail = new LRUNode(0, 0);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    _remove(node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    _add(node) {
        node.next = this.head.next;
        node.prev = this.head;
        this.head.next.prev = node;
        this.head.next = node;
    }

    get(key) {
        if (this.map.has(key)) {
            let node = this.map.get(key);
            this._remove(node);
            this._add(node);
            return node.value;
        }
        return -1;
    }

    put(key, value) {
        if (this.map.has(key)) this._remove(this.map.get(key));
        let node = new LRUNode(key, value);
        this._add(node);
        this.map.set(key, node);
        if (this.map.size > this.capacity) {
            let lru = this.tail.prev;
            this._remove(lru);
            this.map.delete(lru.key);
        }
    }
}
function mainLRU() {
    // Create LRU cache with capacity 2
    const cache = new LRUCache(2);

    cache.put(1, 1);  // Cache: {1=1}
    cache.put(2, 2);  // Cache: {1=1, 2=2}
    console.log(cache.get(1)); // returns 1; Cache order updated to {2=2, 1=1}

    cache.put(3, 3);  // Evicts key 2; Cache: {1=1, 3=3}
    console.log(cache.get(2)); // returns -1 (not found)

    cache.put(4, 4);  // Evicts key 1; Cache: {3=3, 4=4}
    console.log(cache.get(1)); // returns -1 (not found)
    console.log(cache.get(3)); // returns 3
    console.log(cache.get(4)); // returns 4
}

mainLRU();

File: Python/lrucache.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = self.next = None

class LRUCache:
    def __init__(self, capacity):
        self.cap = capacity
        self.cache = {}
        self.head = Node(0, 0)  # Dummy head
        self.tail = Node(0, 0)  # Dummy tail
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node):
        prev, nxt = node.prev, node.next
        prev.next = nxt
        nxt.prev = prev

    def _add(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node

    def get(self, key):
        if key in self.cache:
            self._remove(self.cache[key])
            self._add(self.cache[key])
            return self.cache[key].val
        return -1

    def put(self, key, value):
        if key in self.cache:
            self._remove(self.cache[key])
        node = Node(key, value)
        self._add(node)
        self.cache[key] = node

        if len(self.cache) > self.cap:
            lru = self.tail.prev
            self._remove(lru)
            del self.cache[lru.key]
def main():
    print("LRU Cache Demo")
    cache = LRUCache(2)
    cache.put(1, 1)
    cache.put(2, 2)
    print(cache.get(1))    # returns 1
    cache.put(3, 3)        # evicts key 2
    print(cache.get(2))    # returns -1 (not found)
    cache.put(4, 4)        # evicts key 1
    print(cache.get(1))    # returns -1
    print(cache.get(3))    # returns 3
    print(cache.get(4))    # returns 4

if __name__ == "__main__":
    main()